numpy/core/src/common/simd/avx512/memory.h

```diff
@@ -0,0 +1,94 @@
+    // workaround msvc(32bit) overflow bug, reported at
+    // https://developercommunity.visualstudio.com/content/problem/911872/u.html
+    NPY_FINLINE __m512i npyv__loadl(const __m256i *ptr)
+    {
+        __m256i a = _mm256_loadu_si256(ptr);
+        return _mm512_inserti64x4(_mm512_castsi256_si512(a), a, 0);
+    }
+#else
+    #define npyv__loadl(PTR) \
+        _mm512_castsi256_si512(_mm256_loadu_si256(PTR))
+#endif
+#define NPYV_IMPL_AVX512_MEM_INT(CTYPE, SFX)                                 \
+    NPY_FINLINE npyv_##SFX npyv_load_##SFX(const CTYPE *ptr)                 \
+    { return _mm512_loadu_si512((const __m512i*)ptr); }                      \
+    NPY_FINLINE npyv_##SFX npyv_loada_##SFX(const CTYPE *ptr)                \
+    { return _mm512_load_si512((const __m512i*)ptr); }                       \
+    NPY_FINLINE npyv_##SFX npyv_loads_##SFX(const CTYPE *ptr)                \
+    { return npyv__loads(ptr); }                                             \
+    NPY_FINLINE npyv_##SFX npyv_loadl_##SFX(const CTYPE *ptr)                \
+    { return npyv__loadl((const __m256i *)ptr); }                            \
+    NPY_FINLINE void npyv_store_##SFX(CTYPE *ptr, npyv_##SFX vec)            \
+    { _mm512_storeu_si512((__m512i*)ptr, vec); }                             \
+    NPY_FINLINE void npyv_storea_##SFX(CTYPE *ptr, npyv_##SFX vec)           \
+    { _mm512_store_si512((__m512i*)ptr, vec); }                              \
+    NPY_FINLINE void npyv_stores_##SFX(CTYPE *ptr, npyv_##SFX vec)           \
+    { _mm512_stream_si512((__m512i*)ptr, vec); }                             \
+    NPY_FINLINE void npyv_storel_##SFX(CTYPE *ptr, npyv_##SFX vec)           \
+    { _mm256_storeu_si256((__m256i*)ptr, npyv512_lower_si256(vec)); }        \
+    NPY_FINLINE void npyv_storeh_##SFX(CTYPE *ptr, npyv_##SFX vec)           \
+    { _mm256_storeu_si256((__m256i*)(ptr), npyv512_higher_si256(vec)); }
+
+NPYV_IMPL_AVX512_MEM_INT(npy_uint8,  u8)
+NPYV_IMPL_AVX512_MEM_INT(npy_int8,   s8)
+NPYV_IMPL_AVX512_MEM_INT(npy_uint16, u16)
+NPYV_IMPL_AVX512_MEM_INT(npy_int16,  s16)
+NPYV_IMPL_AVX512_MEM_INT(npy_uint32, u32)
+NPYV_IMPL_AVX512_MEM_INT(npy_int32,  s32)
+NPYV_IMPL_AVX512_MEM_INT(npy_uint64, u64)
+NPYV_IMPL_AVX512_MEM_INT(npy_int64,  s64)
+
+// unaligned load
+#define npyv_load_f32(PTR) _mm512_loadu_ps((const __m512*)(PTR))
+#define npyv_load_f64(PTR) _mm512_loadu_pd((const __m512d*)(PTR))
+// aligned load
+#define npyv_loada_f32(PTR) _mm512_load_ps((const __m512*)(PTR))
+#define npyv_loada_f64(PTR) _mm512_load_pd((const __m512d*)(PTR))
+// load lower part
+#if defined(_MSC_VER) && defined(_M_IX86)
+    #define npyv_loadl_f32(PTR) _mm512_castsi512_ps(npyv__loadl((const __m256i *)(PTR)))
+    #define npyv_loadl_f64(PTR) _mm512_castsi512_pd(npyv__loadl((const __m256i *)(PTR)))
+#else
+    #define npyv_loadl_f32(PTR) _mm512_castps256_ps512(_mm256_loadu_ps(PTR))
+    #define npyv_loadl_f64(PTR) _mm512_castpd256_pd512(_mm256_loadu_pd(PTR))
```

Comment:
The upper half bits are undefined for both these macro's `npyv_loadl_f32 `and `npyv_loadl_f64 ` (for AVX512 and AVX2). I would recommend being consistent with the SSE where its zeros out the upper half bits. (not sure about how its implemented in the ARM and VSX versions).

Reply:
The upper half bits are intentionally left as-is undefined since this intrinsic will mainly use with unpacking and shuffling operations, where no need to touch the upper part.
In case if the upper half bits are required to be filled with zeros then that leave to us two options:

- create a new intrinsic perform this behavior e.g. `npyv_loadlz_*`
- just use combine lower intrinsic `npyv_combinel_*` with zero lanes vector.
  Example:

  ```C
  npyv_f32 a = npyv_loadl_f32(ptr); // {a1, a2, ...}
  a = npyv_combinel_f32(lower_a, npyv_zero_f32()); // {a1, a2, 0, 0}
  ````